src/hotspot/share/gc/parallel/psOldGen.cpp

```diff
@@ -118,8 +118,8 @@ void PSOldGen::initialize_performance_counters() {
 }
 
 HeapWord* PSOldGen::expand_and_allocate(size_t word_size) {
-  assert(SafepointSynchronize::is_at_safepoint(), "precondition");
-  assert(Thread::current()->is_VM_thread(), "precondition");
+  assert(Heap_lock->is_locked(), "precondition");
```

Comment:
Are these methods fine to call from a non safepoint after we have started the GC? If they are not. We should not loosen the precondition this much.

```suggestion
  precond(Heap_lock->is_locked());
  precond(!is_init_completed() || SafepointSynchronize::is_at_safepoint());
  precond(!is_init_completed() || Thread::current()->is_VM_thread());
```

Reply:
While it's true the assert is more loose than it needs to be, if we want the tighter assert here, then we should probably use that tighter assert in all the other places that were loosened too. If we want to do that, could we do that in a follow-up PR instead? I have done pre-integration testing with the object streaming JEP with the changes here, and if we poke around at the asserts again, I'm gonna have to re-run the tests again before integrating.